### DISCUSSION TITLE
audio with meta-tegra

### DISCUSSION THREAD

**Discussion by brgl** [OE4T/tegra-demo-distro#69] [General]
Hi! I was wondering if there were any examples on how to enable HDMI audio with meta-tegra. I admit that alsa is my weak point so I'm struggling to get that working. By default I didn't get any audio output over HDMI. I enabled pulseaudio in my distro features and added alsa-utils. Unfortunately it looks like pulseaudio only installs a user-session systemd service. I don't have other users in my system than root. I tried to modify the systemd service to run pulseaudio system-wide but failed to make it work - aplay tells me:

```
# aplay /usr/share/sounds/alsa/Noise.wav 
W: [pulseaudio] protocol-native.c: Denied access to client with invalid authentication data.
ALSA lib ../../alsa-plugins-1.2.1/pulse/pulse.c:242:(pulse_connect) PulseAudio: Unable to connect: Access denied

aplay: main:828: audio open error: Connection refused
```

Any advice? Should I go without pulseaudio if I only have a single program that will play audio? Any specific settings that are known to work?

**Comment by madisongh** [comment]
Did you try [this PACKAGECONFIG setting](https://github.com/openembedded/openembedded-core/blob/ecb038f14c0b91280ba1532ad94a6ebc64c70644/meta/recipes-multimedia/pulseaudio/pulseaudio.inc#L121)?

**Reply by brgl** [reply]
@madisongh Thanks, I didn't see it before. This does fix the above problem but still no audio over HDMI. When I run aplay with an example file, it exits immediately but without any error code:

```
# time aplay /usr/share/sounds/alsa/Rear_Right.wav 
Playing WAVE '/usr/share/sounds/alsa/Rear_Right.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Mono

real	0m0.064s
user	0m0.020s
sys	0m0.004s
```

`pulseaudio` does get started though.

**Reply by madisongh** [reply]
It's been several years since I configured PA, but I remember that it was difficult to figure out all the pieces and how to configure them. Make sure you've got the ALSA-related PA packages installed and configured. You might also want to check that the tegrahda ALSA sound device is working on its own, without PA.... I do test gstreamer ALSA audio output through HDMI when I run my tests on the dev kits, but if you've got a custom MACHINE setup, you should make sure it's still hooked up properly.

You might find general PulseAudio how-to docs (like the [ArchWiki](https://wiki.archlinux.org/index.php/PulseAudio) page) helpful.

**Reply by brgl** [reply]
@madisongh Do you run your tests on tegra-demo-distro with default settings or some specific configuration? Can you send me the pipeline you're using for testing? I'll start out by replicating your setup and work towards pulseaudio from there.

**Reply by madisongh** [reply]
Yes, I use the demo distro, with the default setup. The scripts for running the tests are included in those builds (the gstreamer-tests package): `/usr/bin/capturetest` for testing camera pipelines, and `/usr/bin/sinteltest` for testing H264 video + audio.

**Reply by brgl** [reply]
@madisongh FYI I managed to configure pulseaudio for the root user. It required me (among common pulseaudio config for running the daemon system-wide as root) to specify the default sink to 0 (which should be the default anyway but isn't) in `client.conf` to make pulseaudio use HDMI. Do you think it's worth adding a small snippet to a .bbappend for pulseaudio in meta-tegra to make sure people don't run into the same problem?

It would look something like this:

```
do_install_append_xavier-nx() {
    sed 's/; default-sink =/default-sink = 0/g' -i ${D}${sysconfdir}/pulse/client.conf
}
```
